src/backend/access/appendonly/appendonly_compaction.c

```diff
@@ -663,13 +663,13 @@ AppendOptimizedDropDeadSegments(Relation aorel, Bitmapset *segnos)
 	/*
 	 * drop segments in batch with concurrent-safety
 	 */
-	LockRelationForExtension(aorel, ExclusiveLock);
+	LockDatabaseObject(RelationGetRelid(aorel), 0, 0, ExclusiveLock);
```

Comment:
```suggestion
	LockDatabaseObject(RelationRelationId, RelationGetRelid(aorel), 0, ExclusiveLock);
```
and the other places too. `LockDatabaseObject` is never used with zero objid.

Reply:
But this is closer to the `LockRelationForExtension` function, which sets the current database's OID in the `locktag_field1` field and the table's OID in the `locktag_field2` field, leaving the `locktag_field3` and `locktag_field4` fields zero.

Reply:
> `LockDatabaseObject` is never used with zero objid.

This means there will be no conflicts with other places!

Reply:
The question is what exact data structure(s) and for which operation(s) we are trying to protect. The comment says:
```
	/*
	 * The algorithm below for choosing a target segment is not concurrent-safe.
	 * Grab a lock to serialize.
	 */
```
So may be we will serialize algorithm by using unique classid  (f.e. ChooseAOSegmentId) and 	`RelationGetRelid(aorel)` as objid?

Reply:
reworked